src/libarena/lib.rs

```diff
@@ -498,21 +454,63 @@ impl<T> TypedArena<T> {
         unsafe {
             let ptr: &mut T = &mut *(self.ptr.get() as *mut T);
             ptr::write(ptr, object);
-            self.ptr.set(self.ptr.get().offset(1));
+            // Purposefully don't use 'ptr.offset' because for arenas
+            // with zero-sized elements this wouldn't change the pointer.
+            self.ptr.set(intrinsics::arith_offset(self.ptr.get(), 1));
```

Comment:
I see. I'll add a test to make sure the right number of zero-sized values is recorded and dropped.